### PR TITLE
Export modules in ir with google protobuf

### PR DESCRIPTION
**Pull request by li-roy**

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -3031,6 +3031,41 @@ def foo3(a):
...
+            self.assertEqual(o, input + torch.ones([2, 2], dtype=torch.float) + m.param + m.param2)
+
+        # test file export
+        import io
+        f = io.BytesIO()
+        torch.onnx._export_module(m_orig, f, OperatorExportTypes.RAW, ExportTypes.ZIP_ARCHIVE)
+        f.seek(0)
+        import zipfile
+        with zipfile.ZipFile(f, 'r', compression=zipfile.ZIP_STORED) as z:
+            self.assertExpected(str([file.filename for file in z.infolist()]))
```
Comment (dzhulgakov) [inline test/test_jit.py]: why not test also __MODEL_PROTO contents?

File: torch/csrc/jit/export.cpp [hunk 2]
```diff
@@ -16,222 +17,177 @@ namespace torch { namespace jit {
...
+#define FAIL_EXPORT(name) \
+      throw std::runtime_error(std::string("ONNX export failed: ") + name + "\n\nGraph we tried to export:\n" + graph->toString());
+    IR_IF(node, PythonOp)
+      auto py_node = static_cast<torch::jit::PythonOp*>(value);
+      FAIL_EXPORT(
+          "Couldn't export Python operator " + py_node->name() +
+          "\n\nDefined at:\n" + getNodeStackTraceString(node))
+    IR_ELSE()
+      // Special error messages for certain types of operators
+      if (node->kind() == aten::expand) {
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: unrelated to this diff - but @jamesr66a - shouldn't it be allowed if we do aten fallback export mode?

File: torch/csrc/jit/export.cpp [hunk 3]
```diff
@@ -16,222 +17,177 @@ namespace torch { namespace jit {
...
+    onnx::ModelProto *model_proto,
+    int64_t onnx_opset_version,
+    onnx_torch::OperatorExportTypes operator_export_type,
+    bool defer_weight_export)
+    : num_blocks_(0),
+      defer_weight_export_(defer_weight_export),
+      operator_export_type_(operator_export_type) {
+  model_proto->set_producer_name("pytorch");
+  model_proto->set_ir_version(3);
+  model_proto->set_producer_version("0.4");
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: maybe do set_doc_string() or change set_producer_name() to signify it's not a real onnx?

File: torch/csrc/jit/export.cpp [hunk 4]
```diff
@@ -16,222 +17,177 @@ namespace torch { namespace jit {
...
+JitEncoder::JitEncoder(
+    onnx::ModelProto *model_proto,
+    int64_t onnx_opset_version,
+    onnx_torch::OperatorExportTypes operator_export_type,
+    bool defer_weight_export)
+    : num_blocks_(0),
+      defer_weight_export_(defer_weight_export),
+      operator_export_type_(operator_export_type) {
+  model_proto->set_producer_name("pytorch");
+  model_proto->set_ir_version(3);
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: there should be constant called IR_VERSION

File: torch/csrc/jit/export.cpp [hunk 5]
```diff
@@ -241,180 +197,546 @@ void encodeBlock(onnx::GraphProto * p_g, Block *b,
...
+      auto g = attr->mutable_g();
+      EncodeGraph(g, node->g(name));
+    } break;
+    case AttributeKind::gs:
+      attr->set_type(onnx::AttributeProto_AttributeType_GRAPHS);
+      for(auto & v : node->gs(name)) {
+        auto g = attr->add_graphs();
+        EncodeGraph(g, v);
+      }
+      break;
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: have also default case? (error)

File: torch/csrc/jit/export.cpp [hunk 6]
```diff
@@ -16,222 +17,177 @@ namespace torch { namespace jit {
...
+      }
+    IR_END()
+#undef FAIL_EXPORT
+  }
+}
 
-void encodeBlock(onnx::GraphProto * p_g, Block *b,
-                const std::vector<at::Tensor> & initializers,
-                ExportContext *ctx, RawDataExportMap* raw_data_export_map);
+class JitEncoder {
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: maybe call it EncoderBase? Jit is kinds of a confusing name (even though it's all over the place)

File: torch/csrc/jit/export.cpp [hunk 7]
```diff
@@ -241,180 +197,546 @@ void encodeBlock(onnx::GraphProto * p_g, Block *b,
...
+  tensor_proto->set_data_type(ATenTypeToOnnxType(tensor.type().scalarType()));
+  auto t = tensor.contiguous().toBackend(at::kCPU);
+
+  // The int64_data field stores offset, and strides
+  tensor_proto->add_int64_data(tensor.storage_offset());
+  for (auto &d : tensor.strides()) {
+    tensor_proto->add_int64_data(d);
+  }
+
+  auto storage_ptr = tensor.data_ptr();
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: hm, I thought data_ptr returns you the shared buffers, but it won't catch the views of tensors (if one of them points to a subset of another). Do we want to support that?

File: torch/csrc/jit/export.cpp [hunk 8]
```diff
@@ -241,180 +197,546 @@ void encodeBlock(onnx::GraphProto * p_g, Block *b,
...
-    validateGraph(graph, operator_export_type);
+std::string idt(size_t indent) {
+  return std::string(indent * indent_multiplier, indent_char);
+}
+
+std::string nlidt(size_t indent) {
+  return std::string("\n") + idt(indent);
+}
+
+void dump(const onnx::TensorProto& tensor, std::ostream& stream) {
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: hm, do we really need the dumping logic now that we don't have nanopb? we could use built in protobuf's TextFormat or DebugString

File: torch/csrc/jit/import.cpp [hunk 9]
```diff
@@ -15,514 +15,308 @@ namespace torch { namespace jit {
...
-
-  return tensor;
 }
 
-Graph_ readSubgraph(const std::string& serialized_subgraph) {
-  pb_istream_t istream = pb_istream_from_buffer(reinterpret_cast<const pb_byte_t *>(serialized_subgraph.data()), serialized_subgraph.size());
-
-  return Reader<Graph_>::read(&istream);
+at::Tensor JitDecoder::buildTensor(const onnx_torch::TensorProto& tensor_proto) {
+  at::Tensor tensor = at::CPU(onnxTypeToATenType(tensor_proto.data_type())).tensor();
```
Comment (dzhulgakov) [inline torch/csrc/jit/import.cpp]: btw, should we store also tensor's device? or things magically turn to cpu after serialization?

File: torch/onnx/utils.py [hunk 10]
```diff
@@ -267,6 +267,27 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
...
+    from torch.onnx.symbolic import _onnx_opset_version
+
+    proto, storage_map = module.export(_onnx_opset_version, operator_export_type)
+
+    if export_type in [ExportTypes.ZIP_ARCHIVE, ExportTypes.COMPRESSED_ZIP_ARCHIVE]:
+        import zipfile
+        compression = zipfile.ZIP_DEFLATED \
+            if export_type == ExportTypes.COMPRESSED_ZIP_ARCHIVE \
+            else zipfile.ZIP_STORED
+        with zipfile.ZipFile(f, 'w', compression=compression) as z:
```
Comment (dzhulgakov) [inline torch/onnx/utils.py]: when we get @jamesr66a's c++ zip integration, if it comes with producing code, we can consider moving it to c++ to avoid an extra copy

File: torch/onnx/utils.py [hunk 11]
```diff
@@ -267,6 +267,27 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
     return torch_out
 
 
+def _export_module(module, f, operator_export_type=OperatorExportTypes.ONNX, export_type=ExportTypes.PROTOBUF_FILE):
```
Comment (dzhulgakov) [inline torch/onnx/utils.py]: we probably should move it outside of onnx/ folder somewhere else

Comment (dzhulgakov) [inline torch/onnx/utils.py]: hm, does it actually support different operator_export_type? from what I understand we don't run the onnx conversion pass here

**Review by dzhulgakov — COMMENTED** [review]
good stuff!

File: torch/csrc/jit/export.cpp [hunk 12]
```diff
@@ -292,85 +239,314 @@ void encodeBlock(onnx::GraphProto * p_g, Block *b,
...
-            getNodeStackTraceString(node));
-      }
+          "\n\nDefined at:\n" + getNodeStackTraceString(node));
     IR_END()
-#undef FAIL_EXPORT
+  }
+  EncodeBlock(attr_proto->mutable_g(), method->graph()->block(), {});
+}
+
+void ModuleEncoder::EncodeTensor(
```
Comment (zdevito) [inline torch/csrc/jit/export.cpp]: Because this isn't virtual-override, I am not sure it ever actually runs.

File: torch/csrc/jit/export.cpp [hunk 13]
```diff
@@ -292,85 +239,314 @@ void encodeBlock(onnx::GraphProto * p_g, Block *b,
...
+  auto storage_ptr = tensor.data_ptr();
+  auto dedup_it = storage_dedup_map_.find(storage_ptr);
+  if (dedup_it != storage_dedup_map_.end()) {
+    tensor_proto->set_doc_string(dedup_it->second);
+  } else {
+    std::string storage_name = "$" + std::to_string(storage_counter_++);
+    tensor_proto->set_doc_string(storage_name);
+    JIT_ASSERT(raw_data_export_map_.count(storage_name) == 0);
+    storage_dedup_map_[storage_ptr] = storage_name;
+    raw_data_export_map_[storage_name] = tensor;
```
Comment (zdevito) [inline torch/csrc/jit/export.cpp]: Careful here about the storages again. `tensor` may not necessarily have a storage pointer that points to the beginning of the storage.

File: torch/csrc/jit/import.cpp (hunk 9 above)
Comment (zdevito) [reply torch/csrc/jit/import.cpp]: I was thinking that everything should turn into a CPU tensor, because loading the module will mmap the storage from the file directly, but maybe it is a better idea to keep all the information and then only use mmap for the CPU tensors.

File: torch/csrc/jit/import.cpp [hunk 14]
```diff
@@ -155,27 +161,162 @@ void buildBlock(const onnx_torch::GraphProto& graph_proto, Block* block,
...
+
+  std::pair<std::shared_ptr<script::Module>, std::string> parseFullName(
+      std::shared_ptr<script::Module> root_module,
+      const std::string fullname);
+
+  const std::unordered_map<std::string, std::string> *storage_export_map_;
+  std::unordered_map<std::string, std::unique_ptr<at::Storage>> storage_map_;
+};
+
+at::Tensor ModuleDecoder::buildParameter(const onnx_torch::TensorProto& tensor_proto) {
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: similar to encoder, we should not duplicate the logic between buildTensor and buildParameter.

File: torch/csrc/jit/import.cpp [hunk 15]
```diff
@@ -155,27 +161,162 @@ void buildBlock(const onnx_torch::GraphProto& graph_proto, Block* block,
...
+      *storage, tensor_proto.int64_data(0), dims, strides);
+  return tensor;
+}
+
+at::Storage* ModuleDecoder::getStorage(at::ScalarType type, std::string name) {
+  auto storage_it = storage_map_.find(name);
+  if (storage_it == storage_map_.end()) {
+    auto storage = at::CPU(type).storage();
+    auto string_it = storage_export_map_->find(name);
+    JIT_ASSERT(string_it != storage_export_map_->end());
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: Rather than assert here, can't we just initialize this entry the first time it is requested? That we there is no "action at a distance"

File: torch/csrc/jit/import.cpp [hunk 16]
```diff
@@ -155,27 +161,162 @@ void buildBlock(const onnx_torch::GraphProto& graph_proto, Block* block,
...
+  std::vector<std::string> vec;
+  std::stringstream ss(fullname);
+  std::string name;
+  while (std::getline(ss, name, '.')) {
+    vec.push_back(name);
+  }
+
+  std::shared_ptr<script::Module> curr = root_module;
+  for (size_t i = 0; i < vec.size() - 1; i++) {
+    if (curr->find_module(vec[i]) != nullptr) {
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: Can simplify:
```
    if (curr->find_module(vec[i]) == nullptr) {
         curr->register_module(vec[i], std::make_shared<script::Module>());
    }
    curr = curr->get_module(vec[i]);
 
```

File: torch/csrc/jit/import.cpp [hunk 17]
```diff
@@ -155,27 +161,162 @@ void buildBlock(const onnx_torch::GraphProto& graph_proto, Block* block,
...
+  storage_map_.clear();
 
-  return graph;
+  for (auto &tensor_proto : graph_proto.initializer()) {
+    std::shared_ptr<script::Module> parent_module;
+    std::string name;
+    std::tie(parent_module, name) = parseFullName(root_module, tensor_proto.name());
+
+    auto param = buildParameter(tensor_proto);
+    parent_module->register_parameter(name, param, tensor_proto.int64_data(1));
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: ```/* is_buffer = */ tensor_proto.int64_data(1) ```
Otherwise it is really confusiong what int64_data(1) is

File: torch/csrc/jit/import.h [hunk 18]
```diff
@@ -1,9 +1,15 @@
 #pragma once
 
 #include "torch/csrc/jit/ir.h"
+#include "torch/csrc/jit/script/module.h"
 
 namespace torch { namespace jit {
 
 TORCH_API std::shared_ptr<Graph> ImportIRGraph(const std::string& serialized_graph, std::vector<at::Tensor> & initializers);
 
+std::shared_ptr<script::Module> ImportIRModule(
```
Comment (zdevito) [inline torch/csrc/jit/import.h]: kind weird to both accept and return root_module.

File: torch/onnx/utils.py (hunk 11 above)
Comment (zdevito) [reply torch/onnx/utils.py]: Yeah, we should not be threading these through for export module because the won't work correctly even if set to ONNX

File: torch/onnx/utils.py (hunk 10 above)
Comment (zdevito) [reply torch/onnx/utils.py]: Everything should move to C++, it would clean up the API a lot since the exporter could also handle writing entries into the zip file, making the function prototypes simpler (no more export maps, e.g.). This is for James' follow up PR.

**Review by zdevito — COMMENTED** [review]
This looks really good! I found a few places where there are bugs (around storage offsets and sharing), and it needs a lot more tests, the contents of which I described in a comment.

File: torch/csrc/jit/import.cpp (hunk 15 above)
Comment (li-roy) [reply torch/csrc/jit/import.cpp]: storage_export_map_ contains all the things that were loaded in. The assert is just a sanity check to make sure that the storage we're looking for is in the set of storages we loaded in.

File: test/test_jit.py (hunk 1 above)
Comment (li-roy) [reply test/test_jit.py]: It's not human readable, so I don't want to do an expects test on it.

File: test/test_jit.py [hunk 19]
```diff
@@ -3031,6 +3031,41 @@ def foo3(a):
...
+
```
[the remark this answers is not in the document]
Comment (li-roy) [reply test/test_jit.py]: I don't think I'm able to even create a ScriptModule with a sparse tensor, so I didn't add that test.

File: torch/csrc/jit/import.cpp [hunk 20]
```diff
@@ -553,20 +222,225 @@ void reconstructOutputTypes(Block *b) {
...
+
+void ModuleDecoder::buildIntermediateValue(Value* value, const std::string& name) {
+  auto it = value_type_map_.find(name);
+  JIT_ASSERT(it != value_type_map_.end());
+  value->setType(buildType(*it->second));
+}
+
+at::Tensor ModuleDecoder::buildParameter(const onnx_torch::TensorProto& tensor_proto) {
+  std::vector<int64_t> strides;
+  std::move(tensor_proto.int64_data().begin() + 3, tensor_proto.int64_data().end(), std::back_inserter(strides));
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: Why are parameter tensors different than other tensors? We should at least have a comment. The +1 and the +3 are completely magical here.

**Review by zdevito — APPROVED** [review]
The changes to encode types look correct. I think this is good to go once we document the encoding of types, and the differences between parameter/tensor encodings.

File: torch/csrc/jit/export.cpp [hunk 21]
```diff
@@ -22,227 +22,176 @@ namespace {
...
-      onnx_type = onnx::TensorProto_DataType_INT64;
-      break;
+      return onnx::TensorProto_DataType_INT64;
     default:
       AT_ERROR("unexpected tensor scalar type");
-      break;
-  }
-  p->set_data_type(onnx_type);
-  // CPU's HalfTensor doesn't have contiguous(), so first calling contiguous()
-  auto t = tensor.contiguous().toBackend(at::kCPU).toType(cast_type);
```
Comment (jamesr66a) [inline torch/csrc/jit/export.cpp]: Why did you remove the type casting? ONNX export for integral types != int32 is now broken
